source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -472,8 +473,12 @@ FilterTrailersStatus Filter::onTrailers(ProcessorState& state, Http::HeaderMap&
     // We would like to process the body in a buffered way, but until now the complete
     // body has not arrived. With the arrival of trailers, we now know that the body
     // has arrived.
-    sendBodyChunk(state, *state.bufferedData(), ProcessorState::CallbackState::BufferedBodyCallback,
-                  false);
+    if (state.bodyMode() == ProcessingMode::BUFFERED) {
+      sendBodyChunk(state, *state.bufferedData(), ProcessorState::CallbackState::BufferedBodyCallback, false);
+    } else {
+      const auto& all_data = state.consolidateStreamedChunks(true);
```

Comment:
do we need to truncate here? 

per the doc of BUFFERED_PARTIAL: "If the body exceeds the configured buffer limit, then the body contents
    // up to the buffer limit will be sent."

I saw it's not truncating at the other place tho.

Reply:
There is no trucating by call consolidateStreamedChunks().  This function basically consolidate all the small chunks, if exists in the chunk_queue_, and send all the data in one gRPC, as you can see here:
https://github.com/envoyproxy/envoy/blob/1987e387c3d60801cb8a6e4c417342ae8a12ed13/source/extensions/filters/http/ext_proc/ext_proc.cc#L427
  
In BUFFERED_PARTIAL mode, if there is data left over in chunk_queue_, then the BUFFER limit should have never been reached since If the data ever exceed the queue limit, no more data will be pushed into the chunk_queue_:
https://github.com/envoyproxy/envoy/blob/1987e387c3d60801cb8a6e4c417342ae8a12ed13/source/extensions/filters/http/ext_proc/ext_proc.cc#L375

Reply:
BTW, ext_proc filter buffer data differently in BUFFERED mode and BUFFERED_PARTIAL mode.

In BUFFERED mode, the data is stored in decoder/encoder callbacks, so state.bufferedData() can retrieve them.
In BUFFERED_PARTIAL mode, the data is copied into chunk elements in the chunk_queue_.

Reply:
Yes, exactly.

Reply:
The code did this for BUFFERED mode, but is missing it for BUFFERED_PARTIAL mode.

Reply:
I mean it will send 2 body RPCs to ext server now.

Reply:
Oh, if the 1st gRPC is already sent and waiting for the response, then more data is received, and now trailer. In this case,  Envoy should store the trailer and returns as here

https://github.com/envoyproxy/envoy/blob/41adbd526871ef5312c1f508d332bb9764ce4499/source/extensions/filters/http/ext_proc/ext_proc.cc#L458

Currently the if check here is missing other non IDLE callback states, so below open PR is to address that issue:

https://github.com/envoyproxy/envoy/pull/29327


But the good point is that we should wait for https://github.com/envoyproxy/envoy/pull/29327 committed before commit this PR.

Reply:
ahh, you should probable stack this PR on that PR, it'd be more clear.

Reply:
Yeah, I found this issue after I sent out PR for that one. Thought about combine these two. But, one is for STREAMED, the other is for BUFFERED_PARTIAL. Maybe it's clearer to keep them separated? I think we can commit the STREAMED mode PR first. 

Converting this PR to draft for now.

Reply:
Actually, this PR and https://github.com/envoyproxy/envoy/pull/29327 covers two completely different issues each corresponding to its own state machine. Combine them will make it very difficult to explain the root cause and fix.

Reply:
@stevenzzzz regarding your original question "should we truncate it?"   No, when trailer arrives, we send the left over data in queue directly.  For BUFFERED_PARTIAL mode, there are basically three cases when trailer arrives:
1) The data is never sent to ext_proc server yet, and there is data in the queue, and not over the limit
2) The data is already sent to ext_proc server, and there is data in the queue, and not over the limit
3) The data is already sent to ext_proc server, and there is data in the queue, and over the limit
No need to truncate in any of these cases.
BTW, it is not possible for the below case in the BUFFERED_PARTIAL mode: 
The data is never sent to ext_proc server yet, and there is data in the queue, and over the limit